spec/src/main/asciidoc/repository.asciidoc

```diff
@@ -112,6 +112,246 @@ Subsequently, an implementation of Jakarta Data, specifically tailored to the ch
+
+- has exactly the same type and name as a persistent field or property of the entity class, or
+- is of type `Limit`, `Pageable`, or `Sort`.
+
+A repository with automatic query methods should be compiled so that parameter names are preserved in the class file (for example, using `javac -parameters`), or the parameter names must be specified explicitly using the `@Param` annotation.
+
+A parameter of an automatic query method may not be annotated with lifecycle annotation.
+
+For example:
+
+[source,java]
+----
+Book bookByIsbn(String isbn);
+
+List<Book> booksByYear(Year year, Sort order, Limit limit)
+----
+
+Automatic query methods _are_ portable between providers.
+
+==== Resource accessor method
+
+A _resource accessor method_ is a method with no parameters which returns a type supported by the Jakarta Data provider.
+The purpose of this method is to provide the program with direct access to the data store.
+
+For example, if the Jakarta Data provider is based on JDBC, the return type might be `java.sql.Connection` or `javax.sql.Datasource`.
+Or, if the Jakarta Data provider is backed by Jakarta Persistence, the return type might be `jakarta.persistence.EntityManager`.
+
+The Jakarta Data implementation automatically recognizes the connection types it supports.
+
+[NOTE]
+A Jakarta Data implementation might allow a resource accessor method to be annotated with additional metadata providing information about the connection.
+
+For example:
+
+[source,java]
+----
+Connection connection();
+
+default void cleanup() {
+    try (Statement s = connection().createStatement()) {
+        s.executeUpdate("truncate table books");
+    }
+}
+----
+
+A repository may have at most one resource accessor method.
+
+==== Additional examples
+
+The following examples demonstrate the use of annotated and automatic query methods with `CrudRepository`.
+
+[source,java]
+----
+@Repository
+public interface ProductRepository extends CrudRepository<Product, Long> {
+  @Query("SELECT p FROM Products p WHERE p.name=?1")  // example in JPQL
+  Optional<Product> findByName(String name);
+}
+----
+
+[source,java]
+----
+@Repository
+public interface ProductRepository extends CrudRepository<Product, Long> {
+  @Query("SELECT p FROM Products p WHERE p.name=:name")  // example in JPQL
+  Optional<Product> findByName(@Param("name") String name);
+}
+----
+
+[source,java]
+----
+@Repository
+public interface ProductRepository extends CrudRepository<Product, Long> {
+
+  // Assumes that the Product entity has attributes: yearProduced
+  List<Product> findMadeIn(int yearProduced, Sort... sorts);
+
+  // Assumes that the Product entity has attributes: name, status.
+  boolean existsWithStatus(String name, Status status);
+
+  // Assumes that the Product entity has attributes: yearProduced
+  void deleteOutdated(int yearProduced);
```

Comment:
These examples don't obey the rules that you defined for automatic queries in that they don't return the entity or multiple entities.  Also, I don't see any way that your proposal distinguishes between exists vs delete vs find.  As I understand from the way you wrote it, it will only do find.  If you want to cover these other things, we would need to include more details from the Query by Parameter Names pattern.

Reply:
Yes, good point, nice catch. This was a pre-existing example that I just moved from where it was before, without proper review.

So the thing is that it looks like you guys had added this language:

> A method name prefix, either `find` or `exists` or `count` or `delete`, specifies the type of query to be performed.

Now, FWIW, I would simply remove that feature.  Neither `exists`, `count`, or `delete` queries are common enough in my experience to be privileged in this way, and, IMO, we should get away from assigning special magical semantics via method naming conventions. It's easy enough to handle these things using `@Query`.

But, again, let's discuss it.

Reply:
Anyway, I fixed the code example to use `@Query`.

Reply:
> Now, FWIW, I would simply remove that feature. Neither `exists`, `count`, or `delete` queries are common enough in my experience to be privileged in this way, and, IMO, we should get away from assigning special magical semantics via method naming conventions. It's easy enough to handle these things using `@Query`.

I don't see those as important either.  I think they were only added based on a misunderstanding that you wanted them.  The user already has alternatives in `@Query` as you mentioned, as well as `existsBy`, `countBy`, and `deleteBy`, all of which provide more function than magical `exists`, `count`, or `delete` based on parameter names.  Your update to switch the examples to `@Query` looks good, so I'll mark this resolved.